### PR TITLE
Improve string parsing and JSON parsing

### PR DESCRIPTION
**Pull request by chqrlie**
- fix JSON parsing of non ASCII string contents
- fix html comment detection at start of source code
- more precise string parsing errors
- more precise JSON parsing errors
- ignore more v8 tests
- improve v8 Failure messages

### CODE REVIEW

File: quickjs.c [hunk 1]
```diff
@@ -18693,22 +18698,19 @@ static __exception int js_parse_string(JSParseState *s, int sep,
...
-                        }
-                        if (do_throw)
-                            js_parse_error(s, "octal escape sequences are not allowed in strict mode");
-                        goto fail;
+                if (c == '0' && !(p[1] >= '0' && p[1] <= '9')) {
+                    /* accept isolated \0 */
+                    p++;
+                    c = '\0';
+                } else
+                if ((c >= '0' && c <= '9')
```
Comment (bnoordhuis) [inline quickjs.c:18706]: I'll never get used to this style :-)

File: quickjs.c [hunk 2]
```diff
@@ -19444,14 +19462,14 @@ static int json_parse_string(JSParseState *s, const uint8_t **pp)
 
     for(;;) {
         if (p >= s->buf_end) {
-            js_parse_error(s, "Unexpected end of JSON input");
-            goto fail;
+            //json_parse_error(s, "Unterminated string");
```
Comment (bnoordhuis) [inline quickjs.c:19465]:
```suggestion
```

File: v8-tweak.js [hunk 3]
```diff
@@ -6,6 +6,21 @@
     globalThis.failWithMessage = function(message) {
         if (count > 99) return
         if (++count > 99) return print("<output elided>")
-        print(String(message).slice(0, 128))
+        print(String(message).slice(0, 300))
+    }
+    globalThis.formatFailureText = function(expectedText, found, name_opt) {
+        var message = "Fail" + "ure";
```
Comment (bnoordhuis) [inline v8-tweak.js]: No semis, please :)

File: v8.js [hunk 4]
```diff
@@ -59,6 +62,16 @@ for (const file of files) {
     }
     // exclude tests that use V8 intrinsics like %OptimizeFunctionOnNextCall
     if (flags["--allow-natives-syntax"]) continue
+    if (flags["--allow-natives-for-differential-fuzzing"]) continue
+    if (flags["--dump-counters"]) continue
+    if (flags["--expose-cputracemark"]) continue
+    if (flags["--omit-default-ctors"]) continue
```
Comment (bnoordhuis) [inline v8.js]: FWIW, I think you can leave this flag out and just blacklist omit-default-ctors-array-iterator.js. All the other tests should either a) just work (no observable difference), or b) come coupled with `--allow-natives-syntax`.

**Review by bnoordhuis — APPROVED** [review]
Sorry, looks like I forgot to actually submit my review. It was still pending.

Mostly LGTM with some comments.

File: quickjs.c (hunk 2 above)
Comment (chqrlie) [reply quickjs.c:19465]: I can remove it, but the error message changed in newer versions of v8...

Comment (bnoordhuis) [reply quickjs.c:19465]: Oh, like that. I wouldn't worry about that, that won't start to be an issue until we upgrade to a newer version of V8's test suite. Dead/commented out code on the other hand bitrots quickly.

File: v8-tweak.js (hunk 3 above)
Comment (chqrlie) [reply v8-tweak.js]: I'll never get used to this style :-)

File: v8.js (hunk 4 above)
Comment (chqrlie) [reply v8.js]: OK, done that.

File: quickjs.c (hunk 2 above)
Comment (chqrlie) [reply quickjs.c:19465]: done.

**Review by bnoordhuis — APPROVED** [review]

**Pull request merged**